homeassistant/components/mqtt/__init__.py

```diff
@@ -57,6 +59,22 @@
 MAX_RECONNECT_WAIT = 300  # seconds
 
 
+# Service call validation schema
+def mqtt_topic(value):
+    """Validate that we can publish using this MQTT topic."""
+    if isinstance(value, str) and all(c not in value for c in '#+\0'):
+        return vol.Length(min=1, max=65535)(value)
+    raise vol.Invalid('Invalid MQTT topic name')
+
+MQTT_PUBLISH_SCHEMA = vol.Schema({
+    vol.Required(ATTR_TOPIC): mqtt_topic,
+    vol.Exclusive(ATTR_PAYLOAD, 'payload'): object,
+    vol.Exclusive(ATTR_PAYLOAD_TEMPLATE, 'payload'): cv.string,
+    vol.Required(ATTR_QOS, default=DEFAULT_QOS): vol.In([0, 1, 2]),
+    vol.Required(ATTR_RETAIN, default=DEFAULT_RETAIN): vol.Coerce(bool),
```

Comment:
The failures are in platform tests, which don't use voluptuous for validation. I added a test that tests ascii qos and retain values and also discovered an empty payload triggered a template expansion of `None` which caused a bus error. Fixing that 'broke' the test for no payload and payload template values because... that still got accepted by the schema and the test only failed because it ran out of memory trying to template None.